### PR TITLE
Rework autons ID mapping generation.

### PR DESCRIPTION
**Pull request by likan999**
This implements the algorithm proposed in
https://github.com/containers/storage/issues/852#issuecomment-798954173,
which is:
1. find available IDs from subuid/subgid file; by subtracting the used
   IDs (from other containers) as well as additional IDs, we get the IDs
   available to allocate;
2. target ID range is [0, requestedSize), subtract the additional IDs;
3. allocate IDs from range in step 1; the number to allocate is the
   number of IDs in step 2;
4. generate a mapping from IDs in step 3 to the ones in step 2.

Closes: https://github.com/containers/storage/issues/852

Signed-off-by: Kan Li <likan@google.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
@giuseppe PTAL

**Comment by ctalledo** [comment]
Hi @likan999, thanks for the PR.

Unfortunately I think this patch does not look correct either. For example, in my host, user `containers` is assigned  subuid range 296608:655360 in `/etc/subuid`. With this patch, crictl + CRIO generate these mappings:

1) "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :

{ContainerID:0 HostID:296608 Size:1}
{ContainerID:1 HostID:296609 Size:65536}      <<< INVALID SIZE (expect 65535)

2) Same as before, but change "run_as_user: 296609":

{ContainerID:0 HostID:296609 Size:1}
{ContainerID:2 HostID:296610 Size:65535}      <<< CONTAINER ID DISCONTINUITY (expect ContainerID 1).

2) Same as before, but change "run_as_user: 296610":

{ContainerID:0 HostID:296610 Size:1}              
{ContainerID:1 HostID:296609 Size:1}              <<< HOST-ID  DISCONTINUITY
{ContainerID:3 HostID:296611 Size:65534}      <<< CONTAINER-ID DISCONTINUITY


Apart from this patch, I still don't quite understand why "run_as_user" refers to the host ID (as opposed to the container ID), as discussed in issue #852. But even if it does, we should return continuous container ID mappings in all cases, ideally mapped to a set of continuous host ID mappings.  Gaps in either the container ID range or host ID range will cause extra complexity to the lower-level runtimes (e.g., fsuid ID mappings, etc.)

**Comment by ctalledo** [comment]
FYI, with top-of-tree master (i.e., without this fix), here is the what I see on the same 3 scenarios in the prior comment:

 1. "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :

{ContainerID:0 HostID:296608 Size:1}
{ContainerID:1 HostID:296610 Size:65535}     <<< HOST-ID DISCONTINUITY  (expect HOST ID 296609) (issue #852)

2. Same as before, but change "run_as_user: 296609":

{ContainerID:0 HostID:296609 Size:1}
{ContainerID:1 HostID:296610 Size:65535}       <<< LOOKS GOOD 

3. Same as before, but change "run_as_user: 296610":

{ContainerID:0 HostID:296610 Size:1}        
{ContainerID:1 HostID:296609 Size:1}                <<<< HOST-ID DISCONTINUITY
{ContainerID:2 HostID:296611 Size:65534}       

Ideally the last set of ranges should have been equivalent to: {ContainerID: 0 HostID: 296610: Size: 65536}

**Comment by likan999** [comment]
> Hi @likan999, thanks for the PR.
> 
> Unfortunately I think this patch does not look correct either. For example, in my host, user `containers` is assigned subuid range 296608:655360 in `/etc/subuid`. With this patch, crictl + CRIO generate these mappings:
> 
> 1. "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :
> 
> {ContainerID:0 HostID:296608 Size:1}
> {ContainerID:1 HostID:296609 Size:65536} <<< INVALID SIZE (expect 65535)
> 
> 1. Same as before, but change "run_as_user: 296609":
> 
> {ContainerID:0 HostID:296609 Size:1}
> {ContainerID:2 HostID:296610 Size:65535} <<< CONTAINER ID DISCONTINUITY (expect ContainerID 1).
> 
> 1. Same as before, but change "run_as_user: 296610":
> 
> {ContainerID:0 HostID:296610 Size:1}
> {ContainerID:1 HostID:296609 Size:1} <<< HOST-ID DISCONTINUITY
> {ContainerID:3 HostID:296611 Size:65534} <<< CONTAINER-ID DISCONTINUITY
> 
> Apart from this patch, I still don't quite understand why "run_as_user" refers to the host ID (as opposed to the container ID), as discussed in issue #852. But even if it does, we should return continuous container ID mappings in all cases, ideally mapped to a set of continuous host ID mappings. Gaps in either the container ID range or host ID range will cause extra complexity to the lower-level runtimes (e.g., fsuid ID mappings, etc.)

Can you give steps to reproduce? At least the unit tests in #853 are passing in this PR.

**Comment by ctalledo** [comment]
> Can you give steps to reproduce? At least the unit tests in #853 are passing in this PR. More specifically, what are the parameters of getAutoUserNS and what should be the expected output? I don't have the crictl at hand so I can't reproduce on my side.

The best way to reproduce is to install crictl and use it in conjunction with CRI-O. I am using the following jsons:

alpine-pod.json:

```json
{                                                                                                                                                                                                                                                                                                                             
    "metadata": {                                                                                                                                                                                                                                                                                                             
        "name": "alpine-sandbox",                                                                                                                                                                                                                                                                                             
        "namespace": "default",                                                                                                                                                                                                                                                                                               
        "attempt": 1,                                                                                                                                                                                                                                                                                                         
        "uid": "pod"                                                                                                                                                                                                                                                                                                   
    },                                                                                                                                                                                                                                                                                                                        
    "annotations": {                                                                                                                                                                                                                                                                                                          
        "io.kubernetes.cri-o.userns-mode": "auto:size=65536"                                                                                                                                                                                                                                                                  
    },                                                                                                                                                                                                                                                                                                                        
    "log_directory": "/tmp",                                                                                                                                                                                                                                                                                                  
    "linux": {                                                                                                                                                                                                                                                                                                                
        "security_context": {                                                                                                                                                                                                                                                                                                 
            "run_as_user": {                                                                                                                                                                                                                                                                                                  
                "value": 296608                                                                                                                                                                                                                                                                                               
            }                                                                                                                                                                                                                                                                                                                 
        }                                                                                                                                                                                                                                                                                                                     
    }                                                                                                                                                                                                                                                                                                                         
} 
```

alpine-container.json

```json
{                                                                                                                                                                                                                                                                                                                             
  "metadata": {                                                                                                                                                                                                                                                                                                               
      "name": "alpine"                                                                                                                                                                                                                                                                                                 
  },                                                                                                                                                                                                                                                                                                                          
  "image":{                                                                                                                                                                                                                                                                                                                   
      "image": "alpine"                                                                                                                                                                                                                                                                                       
  },                                                                                                                                                                                                                                                                                                                          
  "command": [                                                                                                                                                                                                                                                                                                                
      "tail", "-f", "/dev/null"                                                                                                                                                                                                                                                                                               
  ],                                                                                                                                                                                                                                                                                                                          
  "log_path":"container.log",                                                                                                                                                                                                                                                                                                 
  "linux": {                                                                                                                                                                                                                                                                                                                  
  }                                                                                                                                                                                                                                                                                                                           
}     
```

And the following crictl command to start the pod:

```
$ crictl run alpine-container.json alpine-pod.json
```

I then check the mappings that get generated inside CRI-O (e.g., using a the debugger).

**Comment by likan999** [comment]
@ctalledo while I setup the environment, could you test this current version and see if it works (to any potential problems faster)? If it returns incorrect result, could you record the parameters of the function `getAutoUserNSIDMappings`? Thanks.

**Comment by ctalledo** [comment]
> @ctalledo while I setup the environment, could you test this current version and see if it works (to any potential problems faster)? If it returns incorrect result, could you record the parameters of the function `getAutoUserNSIDMappings`? Thanks.

Hi @likan999, sure, would be happy to. Will do this later today and get back to you.

**Comment by ctalledo** [comment]
Hi @likan999,

I tested the latest patch you have for this PR and here are my findings:

1) "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 0" :
{ContainerID:0 HostID:0 Size:1}
{ContainerID:1 HostID:296608 Size:65535}    <<< OK

2) "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :
{ContainerID:0 HostID:0 Size:1}    <<< ERROR (expect HostID 296608)
{ContainerID:1 HostID:296608 Size:65535}    <<< ERROR (expect HostID 296609)


Unfortunately, the debugger is not much help in printing the parameters to getAutoUserNSIDMappings() because some of values are hiding behind several levels of pointers:

```
(dlv) args
size = 65536
availableUIDs = ("*github.com/containers/storage.idSet")(0xc0000105c8)
availableGIDs = ("*github.com/containers/storage.idSet")(0xc000010628)
usedUIDMappings = []github.com/containers/storage/pkg/idtools.IDMap len: 0, cap: 0, nil
usedGIDMappings = []github.com/containers/storage/pkg/idtools.IDMap len: 0, cap: 0, nil
additionalUIDMappings = []github.com/containers/storage/pkg/idtools.IDMap len: 1, cap: 1, [...]
additionalGIDMappings = []github.com/containers/storage/pkg/idtools.IDMap len: 1, cap: 1, [...]
```

It seems like the patch fixes the bug in #852, but it does break "run_as_user" settings for values other than 0. Per @giuseppe, "run_as_user" should refer to the host ID mapping for the container's root user. Thus, whenever "run_as_user = X", we expect a mapping of {ContainerID: 0 HostID: X Size: 1}, followed by other mappings for the remainder of the requested range (e.g., {ContainerID: 1 HostID: X+1 Size: 65535}.

Let me know if I can help in any other way please.

**Comment by likan999** [comment]
> Hi @likan999,
> 
> I tested the latest patch you have for this PR and here are my findings:
> 
> 1. "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 0" :
>    {ContainerID:0 HostID:0 Size:1}
>    {ContainerID:1 HostID:296608 Size:65535}    <<< OK
> 2. "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :
>    {ContainerID:0 HostID:0 Size:1}    <<< ERROR (expect HostID 296608)
>    {ContainerID:1 HostID:296608 Size:65535}    <<< ERROR (expect HostID 296609)
> 
> Unfortunately, the debugger is not much help in printing the parameters to getAutoUserNSIDMappings() because some of values are hiding behind several levels of pointers:
> 
> ```
> (dlv) args
> size = 65536
> availableUIDs = ("*github.com/containers/storage.idSet")(0xc0000105c8)
> availableGIDs = ("*github.com/containers/storage.idSet")(0xc000010628)
> usedUIDMappings = []github.com/containers/storage/pkg/idtools.IDMap len: 0, cap: 0, nil
> usedGIDMappings = []github.com/containers/storage/pkg/idtools.IDMap len: 0, cap: 0, nil
> additionalUIDMappings = []github.com/containers/storage/pkg/idtools.IDMap len: 1, cap: 1, [...]
> additionalGIDMappings = []github.com/containers/storage/pkg/idtools.IDMap len: 1, cap: 1, [...]
> ```
> 
> It seems like the patch fixes the bug in #852, but it breaks "run_as_user" settings for values other than 0. Per @giuseppe, "run_as_user" should refer to the host ID mapping for the container's root user. Thus, whenever "run_as_user = X", we expect a mapping of {ContainerID: 0 HostID: X Size: 1}, followed by other mappings for the remainder of the requested range (e.g., {ContainerID: 1 HostID: X+1 Size: 65535}.
> 
> Let me know if I can help in any other way please.

Thanks @ctalledo , it seems dlv supports pointer dereferencing, could you print contents of the arguments by dereferencing the pointers? Thanks. https://github.com/go-delve/delve/blob/master/Documentation/cli/expr.md

**Comment by ctalledo** [comment]
Hi @likan999,

For the failing case:

```
"io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :

{ContainerID:0 HostID:0 Size:1} <<< ERROR (expect HostID 296608)
{ContainerID:1 HostID:296608 Size:65535} <<< ERROR (expect HostID 296609)
```

The arguments are:

size = 65536
availableUIDs = [{start: 296608, end: 951968}]
usedUIDMappings = []
additionalUIDMappings = {ContainerID: 0, HostID: 0, Size: 1},

**Comment by likan999** [comment]
> Hi @likan999,
> 
> For the failing case:
> 
> ```
> "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :
> 
> {ContainerID:0 HostID:0 Size:1} <<< ERROR (expect HostID 296608)
> {ContainerID:1 HostID:296608 Size:65535} <<< ERROR (expect HostID 296609)
> ```
> 
> The arguments are:
> 
> size = 65536
> availableUIDs = [{start: 296608, end: 951968}]
> usedUIDMappings = []
> additionalUIDMappings = {ContainerID: 0, HostID: 0, Size: 1},

This is interesting. The additionalUIDMappings maps 0 to 0. So is this case broken with HEAD, or is it working on HEAD but broken by this PR? It looks like with HEAD it also maps 0 to 0: https://github.com/containers/storage/blob/9d6cee014f3aa74800ee23c053018a99e5215322/userns.go#L429

@giuseppe @ctalledo Now the questions are:
1. Why is additionalUIDMappings 0->0, instead of 296608->0, in this case?
2. Is this library the one that should handle `run_as_user=X`, or is it the caller of this library who should set up the additionalUIDMappings to 296608->0 based on `run_as_user=X`? I don't see an option that specifies `run_as_user`, so how is this library supposed to know? (By saying "this library" I mean this repo, more specifically, the `CreateContainer` method).
3. Or is `{ContainerID:0 HostID:0 Size:1}` really wrong? In the past, I looked at how my container's user namespace is set up, and I had a vague impression that there are two levels of namespace mappings. Is it possible that in some cases, in an outer namespace it maps X to 0; while in an inner namespace it maps 0 to 0, which is where this library is called?

**Comment by ctalledo** [comment]
Hi @likan999 ,

> Why is additionalUIDMappings 0->0, instead of 296608->0, in this case?

Good point; should be 296608->0. I can't answer since I am not familiar enough with the code.

> Is this library the one that should handle run_as_user=X, or is it the caller of this library who should set up the additionalUIDMappings to 296608->0 based on run_as_user=X? I don't see an option that specifies run_as_user, so how is this library supposed to know? (By saying "this library" I mean this repo, more specifically, the CreateContainer method).

I believe it's the caller of the library (CRI-O in my case), but @giuseppe can confirm.

> Or is {ContainerID:0 HostID:0 Size:1} really wrong? In the past, I looked at how my container's user namespace is set up, and I had a vague impression that there are two levels of namespace mappings. Is it possible that in some cases, in an outer namespace it maps X to 0; while in an inner namespace it maps 0 to 0, which is where this library is called?

No, there is no user-ns nesting in this setup.

**Comment by giuseppe** [comment]
> Hi @likan999 ,
> 
> > Why is additionalUIDMappings 0->0, instead of 296608->0, in this case?
> 
> Good point; should be 296608->0. I can't answer since I am not familiar enough with the code.

yes, when run_as_user it is specified, CRI-O creates the additional mapping

> Something must have changed recently thought (either in this patch or in upstream CRI-O) because in my earlier comment, I indicated than when the pod spec had "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" , I got the following mapping (which is incorrect but does have 296608->0 at least):
> 
> ```
> {ContainerID:0 HostID:296608 Size:1}
> {ContainerID:1 HostID:296609 Size:65536} <<< INVALID SIZE (expect 65535)
> ```
> 
> > Is this library the one that should handle run_as_user=X, or is it the caller of this library who should set up the additionalUIDMappings to 296608->0 based on run_as_user=X? I don't see an option that specifies run_as_user, so how is this library supposed to know? (By saying "this library" I mean this repo, more specifically, the CreateContainer method).
> 
> I believe it's the caller of the library (CRI-O in my case), but @giuseppe can confirm.
> 
> > Or is {ContainerID:0 HostID:0 Size:1} really wrong? In the past, I looked at how my container's user namespace is set up, and I had a vague impression that there are two levels of namespace mappings. Is it possible that in some cases, in an outer namespace it maps X to 0; while in an inner namespace it maps 0 to 0, which is where this library is called?
> 
> No, there is no user-ns nesting in this setup.

If CRI-O creates the mapping 0 -> 0 then that is an issue in CRI-O.  I'll have to verify that.

Were you seeing the same behavior also without re-vendoring this version of containers/storage?

**Comment by giuseppe** [comment]
@ctalledo can you share the .yaml file you've used?

I've tried with:

```
apiVersion: v1
kind: ReplicationController
metadata:
  name: sleep
spec:
  replicas: 1
  selector:
    app: sleep
  template:
    metadata:
      name: sleep
      labels:
        app: sleep
      annotations:
        io.kubernetes.cri-o.userns-mode: "auto:keep-id=true"
    spec:
      securityContext:
        runAsUser: 1002
        runAsGroup: 1003
        fsGroup: 1001
      containers:
      - name: sleep
        command: ["/bin/sh", "-c", "while true; do id; sleep 100000; done"]
        image: alpine
```

and I get both with master and this PR (applied to CRI-O commit: 50b6b5c87598f751468c07fd70d3c3b30fbc9701):

```
# kubectl logs pod/sleep-b8q6s
uid=1002(1002) gid=1003 groups=1001

# cat /proc/158461/uid_map 
         0     265535       1002
      1003     266537      64533
      1002       1002          1

# cat /proc/158461/gid_map 
         0     265534       1001
      1002     266535          1
      1004     266536      64532
      1003       1003          1
      1001       1001          1
```

I've also added this patch to containers/storage:

```
diff --git a/vendor/github.com/containers/storage/store.go b/vendor/github.com/containers/storage/store.go
index 9c08eda69..79b62bf82 100644
--- a/vendor/github.com/containers/storage/store.go
+++ b/vendor/github.com/containers/storage/store.go
@@ -1261,6 +1261,8 @@ func (s *store) CreateContainer(id string, names []string, image, layer, metadat
 
        if options.AutoUserNs {
                var err error
+               fmt.Printf("Request UIDs %v\n", options.AutoUserNsOpts.AdditionalUIDMappings)
+               fmt.Printf("Request GIDs %v\n", options.AutoUserNsOpts.AdditionalGIDMappings)
                options.UIDMap, options.GIDMap, err = s.getAutoUserNS(id, &options.AutoUserNsOpts, cimage)
                if err != nil {
                        return nil, err
```

and I get:
```
Request UIDs [{1002 1002 1}]
Request GIDs [{1001 1001 1}]    
```

Do you get anything different?

**Comment by ctalledo** [comment]
Hi @likan999, @giuseppe ,

My apologies, I think I reported incorrect results in my prior comment. The reason for my mistake is that I had added the "run_as_user" on the crictl's container spec rather than in the pod-spec (I was mislead by [this CRI-O test example](https://github.com/cri-o/cri-o/blob/master/test/testdata/container_config.json)).

Anyway, I re-did the manual testing and things look better now.  Below are the results:

With top of tree CRI-O + containers/storage with this patch:

"io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 0" :
{0 0 1}
{1 296608 65535}    <<<  Better than top-of-tree (starts with host ID 296608 rather than 296609)

"io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :
{0 296608 1}
{1 296609 65535}     <<< Looks good (fixes #852).

"io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296609" :
{0 296609 1}
{1 296608 1}             <<< Legal mapping, but breaks linearity on the host ID
{2 296610 65534}     <<< Looks good

In the last example, the mappings are legal but notice that they are non-linear on the host-ID range. I argue this will cause problems for lower level runtimes (e.g., we use shiftfs in [sysbox-runc](https://github.com/nestybox/sysbox), and shiftfs requires linear mappings, such as {0 296609 65536}).

For comparison, with top of tree CRI-O without this patch:

"io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 0" :
{0 0 1}
{1 296609 65535} 

"io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :
{0 296608 1}
{1 296610 65535}     <<< BUG: issue #852 

"io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296609" :
{0 296609 1}
{1 296610 65535}     <<< Linear mapping

Notice that in the last case, top-of-tree CRIO does generate linear mappings. 

I am not asking that the linearity issue be fixed in this PR, just noting it.

**Comment by likan999** [comment]
> Hi @likan999, @giuseppe ,
> 
> My apologies, I think I reported incorrect results in my prior comment. The reason for my mistake is that I had added the "run_as_user" on the crictl's container spec rather than in the pod-spec (I was mislead by [this CRI-O test example](https://github.com/cri-o/cri-o/blob/master/test/testdata/container_config.json)).
> 
> Anyway, I re-did the manual testing and things look better now. Below are the results:
> 
> With top of tree CRI-O + containers/storage with this patch:
> 
>     1. "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 0" :
>        {0 0 1}
>        {1 296608 65535}    <<<  Better than top-of-tree (starts with host ID 296608 rather than 296609)
> 
>     2. "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :
>        {0 296608 1}
>        {1 296609 65535}     <<< Looks good (fixes #852).
> 
>     3. "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296609" :
>        {0 296609 1}
>        {1 296608 1}             <<< Legal mapping, but breaks linearity on the host ID
>        {2 296610 65534}     <<< Looks good
> 
> 
> In the last example, the mappings are legal but notice that they are non-linear on the host-ID range. I argue this will cause problems for lower level runtimes (e.g., we use shiftfs in [sysbox-runc](https://github.com/nestybox/sysbox), and shiftfs requires linear mappings, such as {0 296609 65536}).
> 
> For comparison, with top of tree CRI-O without this patch:
> 
>     1. "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 0" :
>        {0 0 1}
>        {1 296609 65535}
> 
>     2. "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296608" :
>        {0 296608 1}
>        {1 296610 65535}     <<< BUG: issue #852
> 
>     3. "io.kubernetes.cri-o.userns-mode": "auto:size=65536", "run_as_user: 296609" :
>        {0 296609 1}
>        {1 296610 65535}     <<< Linear mapping
> 
> 
> Notice that in the last case, top-of-tree CRIO does generate linear mappings.
> 
> I am not asking that the linearity issue be fixed in this PR, just noting it.

Thanks for confirming it works. Sure for the 3rd case, we can probably improve the allocation algorithm in the future. Right now it just takes the lowest N available IDs; potentially we can try finding N consecutive IDs. However, this may needs more thoughts, as it is unclear how to define linearity: if the container ids are not contiguous, then is linearity defined as host IDs being contiguous, or they have the same difference to the corresponding container IDs, or something else.

**Comment by ctalledo** [comment]
> Thanks for confirming it works. 

Sure @likan999 , thank you for implementing the fix.

> for the 3rd case, we can probably improve the allocation algorithm in the future. Right now it just takes the lowest N available IDs; potentially we can try finding N consecutive IDs. However, this may needs more thoughts, as it is unclear how to define linearity: if the container ids are not contiguous, then is linearity defined as host IDs being contiguous, or they have the same difference to the corresponding container IDs, or something else.

Yes, the topic of linearity for ID mappings is not something this PR should fix. In my mind, container IDs are always linear, and it would be best if they also map to a linear and consecutive host ID range (e.g., container ID 0 -> 65535 maps to host IDs X + (X+65535)). Otherwise it creates more complexity for the lower level runtimes (shiftfs is one example of something that requires such linearity).

**Comment by kevinlee999** [comment]
@giuseppe mind reviewing this PR?

**Comment by likan999** [comment]
@giuseppe @ctalledo do you mind reviewing the PR?

File: idset.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+	start int
+	// End of the interval (exclusive).
+	end int
+}
+
+func (i interval) length() int {
+	return i.end - i.start
+}
+
+type idSet struct {
```
Comment (ctalledo) [inline idset.go:22]: Suggest adding a brief description of what idSet represents.

File: idset.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+			h = hostIterator()
+		}
+		if c.IsZero() {
+			c = containerIterator()
+		}
+	}
+	return out
+}
+
+func (i interval) Intersect(other intervalset.Interval) intervalset.Interval {
```
Comment (ctalledo) [inline idset.go:164]: Suggestion: group type `interval` and all it's methods on one part of this file, and type `idSet` and all its methods on another part of the file. Currently they are interleaved which makes it harder to read.

**Review by giuseppe — APPROVED** [review]
LGTM

**Review by ctalledo — COMMENTED** [review]
In general this looks good, though I am not super-familiar with the code. In particular, the changes to `getAutoUserNSIDMappings` look good to me. 

I would suggest adding unit tests for the newly introduced `interval` and `idSet` types, given that they each have several methods on which higher level functions rely.

Also, please get a review from a developer that is more familiar with this code that I am (likely @giuseppe).

**Comment by likan999** [comment]
> In general this looks good, though I am not super-familiar with the code. In particular, the changes to `getAutoUserNSIDMappings` look good to me.
> 
> I would suggest adding unit tests for the newly introduced `interval` and `idSet` types, given that they each have several methods on which higher level functions rely.
> 
> Also, please get a review from a developer that is more familiar with this code that I am (likely @giuseppe).

Done addressing the comments: added test cases, moved it around, etc. PTAL.

**Review by ctalledo — APPROVED** [review]
Looks great, thanks for reorganizing the idset.go file and for the writing thorough unit tests for it.

**Comment by rhatdan** [comment]
LGTM

**Pull request merged**